### PR TITLE
feat: add task_history function

### PR DESCRIPTION
**Pull request by ZhiHanZ**
I hereby agree to the terms of the CLA available at: https://docs.databend.com/dev/policies/cla/

## Summary

_Briefly describe what this PR aims to solve. Include background context that will help reviewers understand the purpose of the PR._
Previously, we added task_history table to fetch task execution history. However, given table has limitation, since table level push_down filters are too complex to convert to cloud platform grpc api. and may introduce bugs.
thus far, we need to introduce table function to dedicately convey plausible filter arguments to end user and to elimate possible corner cases.

This pull request contains several features and refactors.
1. Support `task_history` function
2. Support constant-folding on table function named arguments.
3. Update api to contain `root_task_id` as list task execution history filter condition.

### Reference
```sql
TASK_HISTORY(
      [ SCHEDULED_TIME_RANGE_START => <constant_expr> ]
      [, SCHEDULED_TIME_RANGE_END => <constant_expr> ]
      [, RESULT_LIMIT => <integer> ]
      [, TASK_NAME => '<string>' ]
      [, ERROR_ONLY => { TRUE | FALSE } ]
      [, ROOT_TASK_ID => '<string>'] )
```

All arguments are optional

| Argument                      | Description                                                                                                                                                        |
|-------------------------------|--------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| SCHEDULED_TIME_RANGE_START    | The start time of the scheduled time range for which to retrieve task history. This should be a constant expression representing a timestamp.                        |
| SCHEDULED_TIME_RANGE_END      | The end time of the scheduled time range for which to retrieve task history. This should be a constant expression representing a timestamp.                          |
| RESULT_LIMIT                  | An integer specifying the maximum number of task history records to return.                                                                                          |
| TASK_NAME                     | A string representing the name of the task for which to retrieve the history.                                                                                         |
| ERROR_ONLY                    | A boolean value (TRUE or FALSE) indicating whether to retrieve only the task history records that encountered errors.                                                 |
| ROOT_TASK_ID                  | A string representing the root task ID. This is used to filter the history to include only the specified root task and its descendants.                               |

### Example

```sql
-- This query retrieves the first record from the task history where only errors are present.
SELECT * 
FROM task_history(result_limit => 1, error_only => TRUE);

-- This query retrieves the first record from the task history for the task named 'haich'.
SELECT * 
FROM task_history(result_limit => 1, task_name => 'haich');

-- This query retrieves the first record from the task history for the task named 'haich' starting from one day before the current timestamp.
SELECT * 
FROM task_history(result_limit => 1, task_name => 'haich', scheduled_time_range_start => date_add(DAY, -1, CURRENT_TIMESTAMP()));

-- This query retrieves the first record from the task history for the task named 'haich' within the time range from one day before the current timestamp to '2024-06-06'.
SELECT * 
FROM task_history(result_limit => 1, task_name => 'haich', scheduled_time_range_start => date_add(DAY, -1, CURRENT_TIMESTAMP()), scheduled_time_range_end => to_date('2024-06-06'));

-- This query retrieves the first record from the task history for the task named 'haich' within the time range from one day before the current timestamp to '2024-06-06', where only errors are present and filtered by the root task ID 'root'.
SELECT * 
FROM task_history(result_limit => 1, task_name => 'haich', scheduled_time_range_start => date_add(DAY, -1, CURRENT_TIMESTAMP()), scheduled_time_range_end => to_date('2024-06-06'), error_only => TRUE, root_task_id => 'root');

```

- Fixes #[Link the issue here]

## Tests

- [ ] Unit Test
- [x] Logic Test
- [ ] Benchmark Test
- [ ] No Test  - _Explain why_

## Type of change

- [ ] Bug Fix (non-breaking change which fixes an issue)
- [x] New Feature (non-breaking change which adds functionality)
- [ ] Breaking Change (fix or feature that could cause existing functionality not to work as expected)
- [ ] Documentation Update
- [x] Refactoring
- [ ] Performance Improvement
- [ ] Other (please describe):

- - -
This change is [Reviewable](https://reviewable.io/reviews/datafuselabs/databend/15731)

### CODE REVIEW

File: src/query/service/src/table_functions/cloud/task_history.rs [hunk 1]
```diff
@@ -0,0 +1,320 @@
...
+            owners: available_roles
+                .into_iter()
+                .map(|x| x.identity().to_string())
+                .collect(),
+            task_names: vec![],
+            root_task_id: self.args_parsed.root_task_id.clone(),
+            next_page_token: None,
+            page_size: None,
+            previous_page_token: None,
+            task_ids: vec![],
```
Comment (TCeason) [inline src/query/service/src/table_functions/cloud/task_history.rs:178]: Need task_names

File: src/query/service/src/table_functions/cloud/task_history.rs [hunk 2]
```diff
@@ -0,0 +1,320 @@
...
+    ) -> Result<ProcessorPtr> {
+        AsyncSourcer::create(ctx.clone(), output, TaskHistorySource {
+            ctx,
+            args_parsed,
+            is_finished: false,
+        })
+    }
+    async fn build_request(&self) -> Result<ShowTaskRunsRequest> {
+        let tenant = self.ctx.get_tenant();
+        let available_roles = self.ctx.get_available_roles().await?;
```
Comment (TCeason) [inline src/query/service/src/table_functions/cloud/task_history.rs:153]: TODO: https://github.com/datafuselabs/databend/pull/15458/files task privilege need consider this new table function

File: src/query/service/src/table_functions/cloud/task_history.rs [hunk 3]
```diff
@@ -0,0 +1,320 @@
...
+use databend_common_catalog::plan::PushDownInfo;
+use databend_common_catalog::table_args::TableArgs;
+use databend_common_catalog::table_context::TableContext;
+use databend_common_catalog::table_function::TableFunction;
+use databend_common_cloud_control::client_config::build_client_config;
+use databend_common_cloud_control::cloud_api::CloudControlApiProvider;
+use databend_common_cloud_control::pb::ShowTaskRunsRequest;
+use databend_common_config::GlobalConfig;
+use databend_common_exception::ErrorCode;
+pub use databend_common_exception::Result;
```
Comment (flaneur2020) [inline src/query/service/src/table_functions/cloud/task_history.rs:32]: pub use -> use ?

**Comment by flaneur2020** [comment]
LGTM

**Review by TCeason — APPROVED** [review]

**Pull request merged**